File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode* cur = head;
        vector<ListNode*> seen;

        while (cur != nullptr){
            for (int i=0; i<seen.size(); i++){
                if (seen[i] == cur){
                    return true;
                }
            }
            seen.push_back(cur);
            cur = cur->next;
        }

        return false;
    }
};

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return head

        prev = None
        cur = head
        count = 1
        while count < left and cur:
            prev = cur
            cur = cur.next
            count += 1
        
        newHead = cur

        while count < right and cur:
            temp = cur.next.next
            cur.next.next = newHead
            newHead = cur.next
            cur.next = temp
            count += 1
        
        if prev:
            prev.next = newHead
            return head
        else:
            return newHead

        
        